### PR TITLE
Supply a trait to allow user annotations to provide JsonProtocol type hints

### PR DESCRIPTION
**Pull request by davidbiancolin**
This is one sketch of a solution to permit more flexible serialization and deserialization of user annotations defined outside of FIRRTL. Currently in FireSim, we use a hacked up version of the JsonProtocol to do precisely what's included in this PR.

I think it would be very easy to add an extra member to supply a custom serializer/deserializer but Json4s does a pretty good job by itself assuming you give it sufficient type hints. It can for instance handle type-parameterized case classes (see #1207), without needing to write a serializer.

### CODE REVIEW

File: src/main/scala/firrtl/annotations/JsonProtocol.scala [hunk 1]
```diff
@@ -9,6 +9,20 @@ import org.json4s.native.JsonMethods._
...
+  def typeHints(): Seq[Class[_]]
+}
+
+/**
+  * An annotation carrying fully qualified class names for which we'd like the
+  * deserialization protocol to generate type hints for.
+  *
+  * @param typeTags the the additional class names
+  */
+case class DeserializationTypeHintsAnnotation(typeTags: Seq[String]) extends NoTargetAnnotation
```
Comment (azidar) [inline src/main/scala/firrtl/annotations/JsonProtocol.scala]: Is it possible to have the typeTag argument be `Seq[Class[_]]` instead of `Seq[String]`? Or does that defeat the purpose of having this annotation be serializable?

File: src/main/scala/firrtl/annotations/JsonProtocol.scala [hunk 2]
```diff
@@ -74,7 +88,7 @@ object JsonProtocol {
   ))
 
   /** Construct Json formatter for annotations */
-  def jsonFormat(tags: Seq[Class[_ <: Annotation]]) = {
+  def jsonFormat(tags: Seq[Class[_]]) = {
```
Comment (azidar) [inline src/main/scala/firrtl/annotations/JsonProtocol.scala]: I don't fully understand why this change is necessary?

File: src/main/scala/firrtl/annotations/JsonProtocol.scala (hunk 1 above)
Comment (davidbiancolin) [reply src/main/scala/firrtl/annotations/JsonProtocol.scala]: I don't think it's serializable otherwise (but i'd have to check).

File: src/main/scala/firrtl/annotations/JsonProtocol.scala (hunk 2 above)
Comment (davidbiancolin) [reply src/main/scala/firrtl/annotations/JsonProtocol.scala]: The classes for which we want to provide tags  for may no longer be a subclass of Annotation, since there may be arbitrarily nested class instances _within_ the annotation we're trying to serialize.

**Comment by davidbiancolin** [comment]
Naturally this breaks all tests that expected some fixed number of annotations...

File: src/main/scala/firrtl/annotations/JsonProtocol.scala [hunk 3]
```diff
@@ -9,6 +9,20 @@ import org.json4s.native.JsonMethods._
 import org.json4s.native.Serialization
 import org.json4s.native.Serialization.{read, writePretty}
 
+trait HasSerializationHints {
+  // For serialization of complicated constuctor arguments, let the annotation
+  // writer specify additional type hints for relevant classes that might be
+  // contained within
+  def typeHints(): Seq[Class[_]]
```
Comment (jackkoenig) [inline src/main/scala/firrtl/annotations/JsonProtocol.scala]:
```suggestion
  def typeHints: Seq[Class[_]]
```

File: src/main/scala/firrtl/annotations/JsonProtocol.scala (hunk 1 above)
Comment (jackkoenig) [reply src/main/scala/firrtl/annotations/JsonProtocol.scala]: It could be serializable by providing custom serializers!

**Comment by jackkoenig** [comment]
> I think it would be very easy to add an extra member to supply a custom serializer/deserializer but Json4s does a pretty good job by itself assuming you give it sufficient type hints. It can for instance handle type-parameterized case classes (see #1207), without needing to write a serializer.

I didn't realize that Json4s does such a good job, that's cool.

This looks mostly good. I'm not a huge fan of `DeserializationTypeHintsAnnotation` since this provides additional annotations that must be handled.

An alternative (that might not work) could be to put `HasSerializationHints` on the companion object of Annotations, then the JSONProtocol could (reflectively, ew) load the companion object of each annotation class, see if it mixes in that trait, and include the type hints if it does. Thoughts?

**Comment by azidar** [comment]
@Mergifyio backport 1.2.x
